### PR TITLE
[rfc] [ocaml] Enable warning "27" unused variable.

### PR DESCRIPTION
**Pull request by ejgallego**
IMHO this warning is important to be on, specially in the upper layers
that are supposed to handle environments in "let style".

Lots of funny stuff in the diff.

Unfortunately CAMPLX does not understand `[@@@ocaml.warning "-27"]` so
this PR is not ready for merge as `ml4` files will trigger the warning
in abundance due to unused `loc` generated by parsing rules.

Suggestions on how to disable the warning for `ml4`-generated files
are welcome.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
I'm not super comfortable with using `_foo` considering there's no warning for using `_foo`. OTOH if we just use `_` the code gets more confusing.

>Suggestions on how to disable the warning for ml4-generated files
are welcome.

We know which files are GENML4FILES so we can do a specific makefile rule.

**Comment by ejgallego** [comment]
> We know which files are GENML4FILES so we can do a specific makefile rule.

I was afraid this could be too fragile, but I'll give it a try.

> I'm not super comfortable with using _foo considering there's no warning for using _foo. OTOH if we just use _ the code gets more confusing.

My idea was that actually using `_foo` should be forbidden. If you need to use `_foo` then you rename to `foo`.

**Comment by ejgallego** [comment]
I know I have also been using `_bar` for tagging variables "to be removed".

**Comment by SkySkimmer** [comment]
>My idea was that actually using _foo should be forbidden.

That would take a lot of energy to check while reviewing PRs and I think in the end the rule would get violated.

**Comment by ejgallego** [comment]
> That would take a lot of energy to check while reviewing PRs and I think in the end the rule would get violated.

Well that is not worse than the current situation. On the other hand I've never spotted an actual use of `_foo` in the codebase, so maybe humans are not so bad at avoiding that ?

**Comment by SkySkimmer** [comment]
You put some in recently...

Gaëtan Gilbert

**Comment by ejgallego** [comment]
> You put some in recently...

Umm, only as a transient between two PRs, none of them are permanent. I thought the intention as a tag was clear.

**Comment by ejgallego** [comment]
That being said, if we want to be picky we can use `__foo` for tagging "stuff that is supposed to die".

IMHO that circumstantial use of `_evd` as to split a porting patch doesn't constitute evidence.

File: plugins/funind/glob_termops.ml [hunk 1]
```diff
@@ -23,6 +23,61 @@ let mkGHole ()          = DAst.make @@ GHole(Evar_kinds.BinderType Anonymous,Mis
   Some basic functions to decompose glob_constrs
   These are analogous to the ones constrs
 *)
+let glob_decompose_prod =
```
Comment (forestjulien) [inline plugins/funind/glob_termops.ml]: Why do we need to add so many function here ?

File: checker/print.ml [hunk 2]
```diff
@@ -41,10 +41,22 @@ let print_pure_constr fmt csr =
...
+      open_vbox 0;
+      print_string "<"; box_display p; print_string ">";
+      print_cut(); print_string "Case";
+      print_space(); box_display c; print_space (); print_string "of";
+      open_vbox 0;
+      Array.iter (fun x ->  print_cut();  box_display x) bl;
+      close_box();
+      print_cut();
+      print_string "end";
+      close_box()
```
Comment (herbelin) [inline checker/print.ml]: Is this code reworking related?

File: engine/universes.ml [hunk 3]
```diff
@@ -302,6 +292,39 @@ let eq_constr_univs_infer_with kind1 kind2 univs fold m n accu =
...
+	(cstrs := Constraints.add 
+	   (Sorts.univ_of_sort s1, UEq, Sorts.univ_of_sort s2) !cstrs;
+	 true)
+    in
+    let rec eq_constr' m n = 
+      m == n ||	compare_head_gen_proj env eq_universes eq_sorts eq_constr' m n
+    in
+    let res = eq_constr' m n in
+    res, !cstrs
+
```
Comment (herbelin) [inline engine/universes.ml]: Are these changes related?

File: kernel/univ.ml [hunk 4]
```diff
@@ -1026,17 +1026,31 @@ struct
...
+      trivial subtyping relation. *)
+  let from_universe_context univcst freshunivs =
+    let inst = (UContext.instance univcst) in
+    assert (Instance.length freshunivs = Instance.length inst);
+    (univcst, UContext.make (Instance.append inst freshunivs,
+                             create_trivial_subtyping inst freshunivs))
+
+  let subtyping_susbst (_univcst, subtypcst) = 
+      let (ctx, ctx') = (halve_context (UContext.instance subtypcst))in
+      Array.fold_left2 (fun subst l1 l2 -> LMap.add l1 l2 subst) LMap.empty ctx ctx'
```
Comment (herbelin) [inline kernel/univ.ml]: Idem

File: kernel/vm.ml [hunk 5]
```diff
@@ -185,5 +358,34 @@ let apply_whd k whd =
...
+and pr_stack stk =
+  Pp.(match stk with
+      | [] -> str "[]"
+      | s :: stk -> pr_zipper s ++ str " :: " ++ pr_stack stk)
+and pr_zipper z =
+  Pp.(match z with
+  | Zapp args -> str "Zapp(len = " ++ int (nargs args) ++ str ")"
+  | Zfix (_f,args) -> str "Zfix(..., len=" ++ int (nargs args) ++ str ")"
+  | Zswitch _s -> str "Zswitch(...)"
+  | Zproj c -> str "Zproj(" ++ Constant.print c ++ str ")")
```
Comment (herbelin) [inline kernel/vm.ml]: Idem

**Review by herbelin — COMMENTED** [review]
No strong opinion. I somehow feel that a `_` would be "nicer" than a `_foo` in many of the examples but it is not so important.

Considering the size, and how systematic are the changes, this may be better double-checked by a diff in a console than using github.

**Comment by ejgallego** [comment]
Indeed it does seem the rebase didn't go well, sorry for the noise, I didn't mean for the rebase "to go public", but it did, as a side effect of the owners change. My old workflow involved frequent rebases just to keep the PRs in shape, but that didn't imply endorsement.

**Comment by ejgallego** [comment]
Yeah someone the rebase damaged this patch beyond repair [the perils of large-scale refactoring]. So I am closing this.

I do indeed think that after #6857 is implemented adding this warning in a per-component basis will a much more productive approach. For example, it does makes sense we want to be stricter in `coq.kernel`, etc...
